Workspaces Server

Mythicaeda opened the pull request:
* **Tickets addressed:** AERIE-000
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->

- Implements a Workspace Server that stores files inside of a Docker Container
- Updates the DB to remove the User Sequences Table
- Adds Workspace Collaborators
- Refactors the migration script to support non-sql between-migration tasks

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->

## Future work
<!-- What next steps can we anticipate from here, if any? -->

Mythicaeda committed:
DB: Remove User Sequence Table

Mythicaeda committed:
DB: Track known filetypes

Placed in the UI schema as this information is tracked for rendering file contents in the UI.

Mythicaeda committed:
DB: Update Workspace

Mythicaeda committed:
DB: Add workspace collaborators

Mythicaeda committed:
Server: Create Workspace Docker Container

Mythicaeda committed:
DB: Update init_db_users

Grant Sequencing read access on file extension table in the UI schema

Mythicaeda committed:
Server: File and Directory Bindings for Workspace Server

Mythicaeda committed:
Server: Workspace Creation and Deletion Endpoints

Mythicaeda committed:
Test: Add Workspace Server to HealthCheck E2ETest

Mythicaeda committed:
DB: Remove Insert/Delete permissions and update Update for Workspaces

- Insert/Delete is managed by the workspace server
- Add parcel id column and relationship
- Add relationship with actions

joshhaug committed:
Prototype of moving files and directories from one workspace to another

joshhaug committed:
Added copyFile functionality, and refactored to use one method for intra- and inter-workspace copy/move operations

joshhaug committed:
Moving common checks in copy/move ops to be in a utility function

joshhaug committed:
Cleaned up some todos, etc.

joshhaug committed:
Adding additional errors to workspace copy/move operations

Mythicaeda committed:
DB: Update Schema migrations

Now uses integers to track migration ids and tracks information related to "after" tasks in migrations

joshhaug committed:
Adding file-based workspace config stuff to action-server

Mythicaeda committed:
Migration: Update Migration Script to use `print_error`

Mythicaeda committed:
Migration: Refactor code flow for bulk migrations to check for "after" tasks after each step in an up.

Mythicaeda committed:
Migration: Implement "after" tasks for workspaces migration

Mythicaeda committed:
Migration: Second half of SQL Migration

Mythicaeda committed:
Server: Add JWTService

This service decodes the authorization header provided with the request.

Needed for Workspace creation and, eventually, route authorization

Mythicaeda committed:
Migration: First half of SQL migration

Mythicaeda committed:
DB: Update Applied Migrations

Mythicaeda committed:
Server: Add Workspaces Server to Publish Workflow

Mythicaeda committed:
Migration: Update PGCMP to start Workspaces Server

Mythicaeda committed:
Delete User Sequence Tests

dandelany committed:
updates for workspaces service in docker files

joshhaug committed:
Not allowing directory overwrite when handing put operations

joshhaug committed:
Updating codeRunner.ts to use simplified auth

joshhaug committed:
Refactor of placement of WorkspaceFileOpException, and cleanup in codeRunner.ts

joshhaug committed:
Fixed issue with filename not being equal to uploaded file name

joshhaug committed:
Cleanup of isCopyMoveValid function

joshhaug committed:
Moving action-server changes to new branch

joshhaug committed:
Whitespace cleanup

joshhaug committed:
Removed directory case in conflict check

joshhaug committed:
Removed TODO and added check for replacing root directory

joshhaug committed:
Moved copyDirectory and moveDirectory

joshhaug committed:
Adding check back in for filenames being identical

joshhaug committed:
Action server changes to support new aerie-actions library

joshhaug committed:
Update workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java

Co-authored-by: Mythicaeda <theresa.kamerman@jpl.nasa.gov>

joshhaug committed:
Updated comment in workspace bindings, and added override for moveFile method

joshhaug committed:
Updated CopyMoveValid to not include a status

joshhaug committed:
Updated caught conditions in post

joshhaug committed:
Added helpText and cleaned up error messages a bit more for post

dandelany reviewed (changes requested):
One note from testing with @mattdailis earlier today:

We think the API for *creating a directory* should basically ignore the `overwrite` flag and always behave the same way: create the requested directory (or directories, if a nested path is given) *if* they don't exist, otherwise return (don't fail if they do exist). This makes it consistent with how Java `createDirectories` API works - having to pass `overwrite` to get this behavior makes it a bit ambiguous & implies an existing directory would be overwritten by an empty one (which should not happen)

More notes to come as we review code!

dandelany committed:
fixes for improved logging and error handling in workspace service

dandelany committed:
update aerie-actions library to 1.0.0-alpha in action server

dandelany committed:
update workspaces migration numbers to 25/26 since migration 24 was added by rebase

dandelany commented:
@Mythicaeda FYI I rebased and had to bump the migration numbers since we merged the External Events PR. I *think* I updated all of the uses of magic number "24" (in aerie_db_migration.py) but let me know if I missed anything.

dandelany committed:
removed unused override of moveFile, added ternary to skip extra DB call

Mythicaeda committed:
Use "catches" to catch when Workspaces do not exist

Before this, the if statement that was supposed to validate if the workspace existed would throw a NoSuchWorkspaceException in the event that the workspace didn't exist

Mythicaeda committed:
Remove unneeded DB calls when moving/copying within the same workspace

dandelany commented on .github/workflows/publish.yml:
```diff
@@ -69,6 +69,9 @@ jobs:
           - image: aerie-action
             context: action-server
             file: action-server/Dockerfile
+          - image: aerie-workspaces
```
let's make this singular `aerie-workspace` or `aerie_workspace` everywhere (dash for image, _ for container)

dandelany commented on deployment/docker-compose.yml:
```diff
@@ -159,6 +159,26 @@ services:
     networks:
       - aerie_net
+  aerie_workspace:
+    container_name: aerie_workspaces
```
singular here too

dandelany committed:
Merge pull request #1712 from NASA-AMMOS/feat/workspaces-more-improvements

Workspaces copy/move support

mattdailis committed:
Merge pull request #1713 from NASA-AMMOS/feat/action-server-workspace-refactor

Action server changes to support new aerie-actions library

Mythicaeda committed:
Rename "workspace_server" folder to "workspace-server"

Rename makes it match our naming conventions

Mythicaeda committed:
Replace "workspaces" with singular "workspace"

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -0,0 +1,454 @@
+  @Override
+  public void apply(final Javalin javalin) {
+    javalin.routes(() -> {
+      //before("/ws/*", authorize); <- dont want to force auth on the health check.
```
This is commented out, but without it, we don't auth users (except on the `createWorkspace` route). We think this should be:
```
before("/ws/*", this::authorize);
```
does that seem right @Mythicaeda ?

Mythicaeda replied:
Oh, it's commented out because there was no implemented `authorize` method when I drafted it. I'll uncomment it, just be aware that it's currently only checking that an unexpired, valid JWT was provided ("valid" meaning the user can access the role they're using and that the signature is correct), as opposed to a detailed permissions check (ie, "viewer" role can't use the `edit` endpoints).

mattdailis commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -0,0 +1,454 @@
+
+      workspaceLocation = Path.of(bodyJson.getString("workspaceLocation"));
+      workspaceName = bodyJson.containsKey("workspaceName") ? bodyJson.getString("workspaceName") : workspaceLocation.toString();
+    } catch (NullPointerException npe) {
+      context.status(400).result("Mandatory body parameter is null. Request body format is:\n" + helpText);
+      return;
+    } catch (JsonException je) {
```
I'm curious - which line of code above is expected to sometimes throw a null pointer exception? Would it be possible to do some kind of null check instead of catching NullPointerException?

In general I'm of the opinion that catching NPE is something to be avoided because it could easily mask a real bug

Mythicaeda replied:
L111-L112: 
```java
parcelId = bodyJson.getInt("parcelId");
final var workspaceString = bodyJson.getString("workspaceLocation");
```

When you call `json.get('key')` on a key not in the json object, the method throws an NPE. Since the code needed a `try-catch` anyway to catch JSON parsing errors (such as if `parcelId` was set to a `string`), I decided to catch the thrown exception rather than set up `if` statements to "pre-catch" it.

> In general I'm of the opinion that catching NPE is something to be avoided because it could easily mask a real bug

While I agree with this in general, if the code was to _not_ catch the NPE here, all the user would get in the event that their misformatted body caused an NPE would be a generic `500` status, which is not useful information for them.

mattdailis replied:
> Since the code needed a try-catch anyway to catch JSON parsing errors (such as if parcelId was set to a string), I decided to catch the thrown exception rather than set up if statements to "pre-catch" it.

Thanks, that does make sense, it does make the code easier to write. As a reader, here's what makes me nervous about seeing "catch NullPointerException" anywhere in a program:

These are the possible sources of a null pointer exception in this try catch block:
- reader.readObject() (if reader is null)
- bodyJson.getInt("parcelId") (if bodyJson is null or if the key is missing)
- bodyJson.getString("workspaceLocation")  (if bodyJson is null or if the key is missing)
- workspaceString.contains("/") (if workspaceString is null)
- context.status(400) (if context is null)
- bodyJson.getString("workspaceLocation") (if bodyJson is null or if the key is missing)
- bodyJson.containsKey("workspaceName") (if bodyJson is null or if the key is missing)
- bodyJson.getString("workspaceName") (if bodyJson is null or if the key is missing)
- workspaceLocation.toString() (if workspaceLocation is null)

It's a little overwhelming as a reader of this code to try to figure out where it's expected to come from. This argument generalizes to catching any sort of "unchecked" exception, but null pointer exceptions are particularly insidious because they can come from absolutely anywhere.

From a user perspective, all of the above failures get turned into the same message, which I think misses out on an opportunity to give a specific error message. From a debugging perspective, if there is a bug, there is no stack trace to refer to to figure out which line of code caused the error.

> set up if statements to "pre-catch" it.

I took a couple cracks at it, I'm not sure either of these is pretty, but I do think they give better error messages and don't mask potential bugs:
- https://github.com/NASA-AMMOS/aerie/commit/78cc626b98390c0cb772f3eebed8136fc755f5e8
- https://github.com/NASA-AMMOS/aerie/commit/9466db70f857129b6c0caba1046882a5312b6b38

Any preference between those two, or other ideas we should consider?

Mythicaeda replied:
If you revert the changes done to L120 `workspaceName` and return it to its ternary statement (use if provided, else use the location), I prefer the second one.

Though, please, let me push my pending changes first. If I had realized you guys were going to commit to my branch, I would've pushed last night.

Mythicaeda replied:
Actually, correction, you used the wrong method in the second version (you wanted to do `containsKey`, you did `get`, which will throw an NPE if it's absent), I'll just fix it and commit it, I know what was intended

mattdailis replied:
Sounds good - I have no intention of committing to this branch before consulting you first! I made new branches for those two commits just for demonstration purposes (EDIT: whoops! didn't see Dan's commits from yesterday. I stand corrected)

> you did get, which will throw an NPE if it's absent

Is that true? My read on the code was that it acts like `Map`, which returns `null` if the key is missing https://github.com/javaee/jsonp/blob/4ee2eded959188ff0948950a1bd8cd7b07563310/impl/src/main/java/org/glassfish/json/JsonObjectBuilderImpl.java#L340-L343

I do think `containsKey` is clearer though :+1:

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/JWTService.java:
```diff
@@ -0,0 +1,80 @@
+      case "HS256" -> Algorithm.HMAC256(key);
+      case "HS384" -> Algorithm.HMAC384(key);
+      case "HS512" -> Algorithm.HMAC512(key);
+      default -> Algorithm.none();
```
To be defensive, we should `throw` if the `typeString` is not a known algorithm - as is, a user could make a typo eg `hs_256` in their config and it would "fail open" - ie. it would work but would silently not authenticate tokens. We could (if needed) also have a "none" case here for explicitly using no algorithm

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -0,0 +1,312 @@
+
+    for(final var f : contents) {
+      if(Files.isSymbolicLink(f.toPath()) || !f.isDirectory()) {
+        success = success && rm(f);
```
The doc comment for this implies that it will try to delete as much as possible, not stopping on failures, but it seems like this `success && ...` pattern will short-circuit and stop on failures, is this intended?

Mythicaeda replied:
Fixed it. It should not not short-circuit

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -0,0 +1,312 @@
+  }
+
+  @Override
+  public boolean deleteWorkspace(final int workspaceId) throws NoSuchWorkspaceException, SQLException {
```
minor: in general it's a bit confusing to have functions which sometimes `throw` on failure and sometimes `return false` depending on how they fail - would be nice to standardize and maybe always use exceptions. I think the `Files` API does a better job at this than the `File` API

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -0,0 +1,312 @@
+      walkList.removeFirst(); // remove the initial path
+      return new DirectoryTree(path, walkList, postgresRepository.getExtensionMapping());
+    } catch (IOException io) {
+      return null;
```
We should probably throw this Exception up to the caller to get the message - as-is any IOException will be reported as "directory does not exist" but there could be other reasons for the failure

Mythicaeda replied:
Exception is now bubbled-up

dandelany commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/DirectoryTree.java:
```diff
@@ -0,0 +1,96 @@
+    }
+  }
+
+  private static class FileNode {
```
Minor/later thing, but a lot of this file helper stuff seems like the kind of thing that might be better to pull from a library & reduce how much we need to maintain. Might look into using the Apache `FileUtils` library later & see how much that buys us.

Mythicaeda replied:
Added doc comments to explain why this is not using Apache `FileUtils`

dandelany committed:
improvements to JWTService for error-handling and edge cases

dandelany committed:
add safe resolveSubPath util to workspace service, + tests

dandelany committed:
safer path resolution for metadata files in workspace service

dandelany reviewed (changes requested):
@mattdailis and I did some review on this today, notes above. The Java service is pretty well covered, but we still need to review the migration script (tomorrow). I also pushed some additional fixes to this branch since we're time-constrained - @Mythicaeda please take a look and let me know if you have any issues with these:

* c5f3ee1d7efd861901dc0ad20625ffca6ed71af4 ensures that we don't use `Algorithm.none` by accident for JWTs - if there is a good reason to support it we can add it back as an explicit "none" option, but it shouldn't be the default.
* 00c558e8e615e08f8bb33b971faa99c8c7fb3f95 adds and uses a utility for safely resolving subpaths within workspaces. Without this, we are vulnerable to path traversal attacks, eg. a user could access/modify a relative path outside their workspace like `../otherworkspace/someones.file` or even an absolute path like `/etc/passwd` (because `Path.of("/foo").resolve(Path.of("/etc/passwd")) == "/etc/passwd"`). Also added unit tests for this - lmk if you'd prefer this to be in a separate util file.
* 58e68c9dffc0e323c5fdee570ab4004606725c0b adds slightly more path safety by ensuring metadata files are siblings in case one of the file extension values has `..`/other unsafe things.

There might be a few more of these path-safety-type issues remaining, let me know if you see any.

Mythicaeda committed:
Replace NPE catch with `containsKey` check

Mythicaeda committed:
Add `before` handler for workspace routes

Mythicaeda committed:
Reject path routing characters from Workspace Root Dir Name

Mythicaeda committed:
Fix bug where move and delete were fast failing

Mythicaeda committed:
Bubble up and catch IOException when listing directory contents

dandelany committed:
workspace service - skip auth for browser preflight requests

dandelany commented on deployment/hasura/migrations/Aerie/25_workspaces_setup/up.sql:
```diff
@@ -0,0 +1,224 @@
+    set workspace_id = iws.new_wid
+  from inserted_ws iws, parcels_per_ws ppw, sequencing.workspace ws
+  where us.workspace_id = ws.id
+    and iws.name = ws.name || ' (' || ppw.pname ||')'
```
minor: @Mythicaeda can you clarify if this line is needed or not? Was trying to understand if it was safe to remove & just match on `id`s instead

dandelany replied:
nevermind, we understand why this is needed, but would prefer to keep the old workspace ID in `inserted_ws` & then do a match like `and iws.old_workspace_id = ws.id` here to avoid any string/name matching brittleness

dandelany replied:
on third thought... it doesn't seem like `parcel.name` is guaranteed to be unique, so this may be more of a requirement than a preference, since *not* using IDs here could result in a name collision?

Mythicaeda replied:
The update on L199 fixes any workspace name conflicts.

Mythicaeda replied:
> nevermind, we understand why this is needed, but would prefer to keep the old workspace ID in inserted_ws & then do a match like and iws.old_workspace_id = ws.id here to avoid any string/name matching brittleness

This is impossible; you cannot return an input field from an `insert` statement. Trust me, that was my first approach. I do see the concern about the string matching, however

Mythicaeda committed:
Use a temp id column to avoid string matching in migration

Mythicaeda committed:
Elaborate on saveFile comment in db_migration script

dandelany commented on workspace-server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceAppDriver.java:
```diff
@@ -0,0 +1,125 @@
+    });
+
+    // Start the HTTP server.
+    javalin.start(9999);
```
can we change this to `javalin.start(configuration.httpPort());` or even `javalin.start();` here since the connector already defines the port? `9999` seems like it will be overridden/not necessary?

Mythicaeda replied:
I changed it to the first option to align with `AerieAppDriver`

Mythicaeda committed:
Add class-level doc comments to DirectoryTree

- Validate that `root` is actually always a directory

Mythicaeda committed:
Ban use of 'overwrite' flag when creating a directory

Mythicaeda committed:
Update `javalin.start` for Workspace Server

Mythicaeda committed:
Fix bugs in move/copy

Two bugs:
1) Fix bug where the user could attempt to move or copy a folder into itself
2) Fix bug where server always accidentally reported a 200 status, regardless of status set during move and copy

mattdailis committed:
Check for admin secret in workspace server

mattdailis committed:
Update migration script to fix status and provide user id

dandelany committed:
add message if migration has pending tasks to run

dandelany committed:
pass correct admin headers to workspace service in __apply_workspaces_migration__

dandelany committed:
Improved handling of error messages for Unauthorized results in Workspace service

dandelany committed:
pass correct flags to hasura metadata commands in the migration script

dandelany committed:
add HASURA_GRAPHQL_ADMIN_SECRET to docker compose

dandelany committed:
small fixes to migration script for admin DB user

mattdailis committed:
Remove .user suffix and change filename uniqueness strategy

dandelany committed:
add log statement when done with apply workspace migration

Mythicaeda committed:
Change DirectoryTree to use sorted TreeMap instead of unsorted HashMap

dandelany committed:
update to migration script to correctly give unique names to workspace files

dandelany committed:
Merge pull request #1719 from NASA-AMMOS/feat/workspace-improvements-admin-secret

Feat/workspace improvements admin secret

dandelany committed:
update aerie-actions library to 1.0.0

dandelany committed:
Merge branch 'develop' into feat/workspaces-improvements

dandelany reviewed (approved):
#1719 has been merged into this branch, now thoroughly tested (including migrations) and approved. I also updated the `aerie-actions` dependency to `1.0.0` now that it has been published. The security scan should pass on `develop` after the merge. 

Merging this for testing on aerie-dev & release 3.6.0.  Thanks everyone for the hard work on this!
